presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/TestMemo.java

```diff
@@ -0,0 +1,265 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.sql.planner.iterative;
+
+import com.facebook.presto.sql.planner.PlanNodeIdAllocator;
+import com.facebook.presto.sql.planner.Symbol;
+import com.facebook.presto.sql.planner.plan.PlanNode;
+import com.facebook.presto.sql.planner.plan.PlanNodeId;
+import com.google.common.collect.ImmutableList;
+import org.testng.annotations.Test;
+
+import java.util.List;
+
+import static org.testng.Assert.assertEquals;
+
+public class TestMemo
+{
+    private PlanNodeIdAllocator idAllocator = new PlanNodeIdAllocator();
+
+    @Test
+    public void testInitialization()
+            throws Exception
+    {
+        PlanNode plan = node(node());
+        Memo memo = new Memo(idAllocator, plan);
+
+        assertEquals(memo.getGroupCount(), 2);
+        assertMatchesStructure(plan, memo.extract());
+    }
+
+    /*
+      From: X -> Y  -> Z
+      To:   X -> Y' -> Z'
+     */
+    @Test
+    public void testReplaceSubtree()
+            throws Exception
+    {
+        PlanNode plan = node(node(node()));
+
+        Memo memo = new Memo(idAllocator, plan);
+        assertEquals(memo.getGroupCount(), 3);
+
+        // replace child of root node with subtree
+        PlanNode transformed = node(node());
+        memo.replace(getChildGroup(memo, memo.getRootGroup()), transformed, "rule");
+        assertEquals(memo.getGroupCount(), 3);
+        assertMatchesStructure(memo.extract(), node(plan.getId(), transformed));
+    }
+
+    /*
+      From: X -> Y  -> Z  -> W
+      To:   X -> Y' -> Z' -> W
+     */
+    @Test
+    public void testReplaceNonLeafSubtree()
+            throws Exception
+    {
+        PlanNode w = node();
+        PlanNode z = node(w);
+        PlanNode y = node(z);
+        PlanNode x = node(y);
+
+        Memo memo = new Memo(idAllocator, x);
+
+        assertEquals(memo.getGroupCount(), 4);
```

Comment:
why not to use `IterativeOptimizer` here with some simple lambda rule? Then you could also check the traversal of `IterativeOptimizer`

Something like:
```
optimizer(node ->  if (node == y || node == z) return node(node.getSource())).optimize(x)
```